Faster version of compareFilenames()

redmunds opened the pull request:
The `compareFilenames()` function can sort files by either name first (then extension) or extension first (and then name). Old code executes _both_ comparisons and then only uses the results that it needs. The vast majority of comparisons only need either name comparison or extension comparison (not _both_).

This code now breaks the 2 search types into sub-functions which has these 2 benefits:
1. Only execute comparisons needed
2. `var` storage is function scope, so it's not allocated unless it's used.

For file tree (name-first) comparison, my measurements on Win7 indicate that this function uses **~40% less time**.

redmunds committed:
faster version of compareFilenames()

MarcelGerber commented:
Wouldn't it be even faster if we applied `.toLocaleLowerCase` to the whole filename first, so we don't have to do it twice in the case where both checks are performed?

redmunds commented:
> Wouldn't it be even faster if we applied `.toLocaleLowerCase` to the whole filename first, so we don't have to do it twice in the case where both checks are performed?

If new vars are created for the lowercase filenames then I suspect that it could opposite effect since the it's not common do need both checks and this would cause more storage to be allocated. But re-using existing `filename1` and `filename2` vars might work. This needs to be profiled.

MiguelCastillo commented on src/file/FileUtils.js:
```diff
@@ -434,19 +434,24 @@ define(function (require, exports, module) {
+        }
+        
+        function cmpNames() {
+            if (brackets.platform === "win") {
```
What do you think about memoizing this?  Not really trying to micro optimize.  Just how I would do it; I generally try to reduce conditionals in my main execution branch when I can.

redmunds replied:
Memoizing helps when a function is called many times with same parameter vales. During a call to `.sort()` the same 2 *fullpaths* should never be compared more than once. This function compares just the name partion of the fullpath, so common names such as **index.html** will be compared more than once, but I think this is an exception, So the initial call will have the overhead of memoizing but very little of the benefits. It should help on subsequent calls to `.sort()`. But, when sorting all of the file in a file system, the combination of parameter values is huge, so I am not sure memoization is worth the overhead. Once again, it needs to be profiled.

redmunds replied:
If you're interested in another way to solve the file system sorting problem, take a look at the `getSortedFiles()` function I wrote in the original PR [here](https://github.com/adobe/brackets/pull/11128/files#diff-0ba586b1d4dc4a89d8405edd63a72644R371).

To sort fullpaths, `comparePaths()` splits each path into an array of folders with last item being filename. This is a good technique, but it *splits both paths for every, single fullpath comparison*. In `getSortFiles()`, it splits paths all of the file before sorting starts (so it only happens once for each file) and then executes the sort. This seems to reduce sorting time by another ~16%.

I did not suggest suggest this to be taken for Brackets because it only handles the sort by filename case, not file extension required by the working set, and it's currently not called anywhere in master, so it would need to be factored in somewhere.

MiguelCastillo replied:
Sorry I didn't have a different term. But the only part I was saying to cache is the if statement where we are checking for the platform...  We don't need that check every time we run the code.

``` javascript
var _cmpNames = (function() {
   if (brackets.platform === 'window') {
     return  windows specific code;
   }
   else {
      return what we need;
   }
})();

function compareFilenames(filename1, filename2, extFirst) {
   function cmpNames() {
      return _cmpNames();
   }
}
```


Or some abstraction of this :)

redmunds replied:
Ah. Good idea!

redmunds committed:
only lowercase string once

redmunds committed:
define os-specific function to compare file names

redmunds commented:
@MarcelGerber I implemented "only lowercase once" in commit https://github.com/adobe/brackets/commit/9b3856a9b628955fdb51c568d27c857374038c4a

@MiguelCastillo I implemented "define os-specific function to compare file names" in commit https://github.com/adobe/brackets/commit/61c1c9975484fde04e20d5ac9baff6425d1c0913

I measured performance by putting a timer in `_sortDirectoryContents()` function in FileTreeView.js. Time reduction compared to master is now ~50%

redmunds committed:
code cleanup

TomMalbran commented on src/file/FileUtils.js:
```diff
@@ -426,29 +426,53 @@ define(function (require, exports, module) {
-            lang   = brackets.getLocale(),
-            cmpExt = ext1.toLocaleLowerCase().localeCompare(ext2.toLocaleLowerCase(), lang, {numeric: true}),
-            cmpNames;
+        var lang = brackets.getLocale();
```
Would it be faster if lang could be a fourth optional parameter, and we can use it like `lang = lang || brackets.getLocale()`, so that it would be possible to call that function once, and not on every compare?

redmunds replied:
@TomMalbran Yes. I profiled that at one time and it didn't seem like much of a savings, but everything counts, so I'll take another look at that.

redmunds replied:
@TomMalbran I finally got around to looking at this again. Adding a `locale` parameter is easy, but it only get passed when this function is called directly. This function is usually called (either directly or via `comparePaths()`) as a standard sort function (only first 2 params are passed). The exceptions are:
* Working Set sorting which is only a small set of files
* "Directory First" sorting which only benefits Windows (or anyone who turns on that pref)

So it doesn't seems like it offers enough benefit to make this change.

MiguelCastillo commented on src/file/FileUtils.js:
```diff
@@ -426,29 +426,53 @@ define(function (require, exports, module) {
-            filename1 = getFilenameWithoutExtension(filename1);
-            filename2 = getFilenameWithoutExtension(filename2);
+        filename1 = filename1.toLocaleLowerCase();
+        filename2 = filename2.toLocaleLowerCase();
```
@redmunds something I am sure you are well aware of, but I will just go ahead and mention it anyways.  Is that some systems (a.k.a linux) have case sensitive file systems by default...  So converting file paths to lower is actually breaking behavior. I don't think I would convert these file names to lower case.

redmunds replied:
@MiguelCastillo That was already being done -- it's just more efficient now :)

The problem that this fixes is sorting like this:

```
A
B
C
a
b
c
```

Which is annoying even for case-sensitive systems. It should be like:

```
A
a
B
b
C
c
```

I don't think anyone has names that are identical except for case anyway.

MiguelCastillo replied:
Oh cool.  Thank you for explaining. I should have read the code a lil more :)

redmunds commented:
@abose This PR is ready for as final review.

MarcelGerber commented:
Looks good. Merging.

abose commented:
Thanks @redmunds .

redmunds commented:
Thanks everyone who helped with this one.
